src/github.com/matrix-org/dendrite/clientapi/writers/createroom.go

```diff
@@ -117,12 +130,147 @@ func createRoom(req *http.Request, roomID string) util.JSONResponse {
 	// This differs from Synapse slightly. Synapse would vary the ordering of 3-7
 	// depending on if those events were in "initial_state" or not. This made it
 	// harder to reason about, hence sticking to a strict static ordering.
+	// TODO: Synapse has txn/token ID on each event. Do we need to do this here?
+	eventsToMake := []fledglingEvent{
+		{"m.room.create", emptyString(), common.CreateContent{Creator: userID}},
+		{"m.room.member", &userID, common.MemberContent{Membership: "join"}}, // TODO: Set avatar_url / displayname
+		{"m.room.power_levels", emptyString(), common.InitialPowerLevelsContent(userID)},
+		// TODO: m.room.canonical_alias
+		{"m.room.join_rules", emptyString(), common.JoinRulesContent{"public"}},                 // FIXME: Allow this to be changed
+		{"m.room.history_visibility", emptyString(), common.HistoryVisibilityContent{"joined"}}, // FIXME: Allow this to be changed
+		// TODO: m.room.guest_access
+		// TODO: Other initial state items
+		// TODO: m.room.name
+		// TODO: m.room.topic
+		// TODO: invite events
+		// TODO: 3pid invite events
+		// TODO m.room.aliases
+	}
+
+	for i, e := range eventsToMake {
+		depth := i + 1 // depth starts at 1
+
+		builder := gomatrixserverlib.EventBuilder{
+			Sender:   userID,
+			RoomID:   roomID,
+			Type:     e.Type,
+			StateKey: e.StateKey,
+			Depth:    int64(depth),
+		}
+		builder.SetContent(e.Content)
+		ev, err := buildEvent(&builder, builtEventMap, cfg)
+		if err != nil {
+			return util.ErrorResponse(err)
+		}
+		builtEventMap[common.StateTuple{e.Type, *e.StateKey}] = ev
+		builtEvents = append(builtEvents, ev)
+	}
+	authEvents := authEventProvider{builtEventMap}
+
+	// auth each event in turn
+	for _, e := range builtEvents {
+		if err := gomatrixserverlib.Allowed(*e, &authEvents); err != nil {
```

Comment:
You need to check the whether the event is allowed by the state at the event. If you move the check to inside the first loop then it should be more correct.